geometry/meshop.cpp

```diff
@@ -252,52 +256,92 @@ void saveAsPly( const Mesh &mesh, const boost::filesystem::path &filepath){
-                << "Can't open " << filename;
-    }
-
-    f.exceptions(std::ios::badbit | std::ios::failbit);
-
-    // read header
-    std::string line;
-    int nvert = -1, ntris = -1;
-    do {
-        if (getline(f, line).eof()) break;
-        boost::algorithm::trim(line);
-        sscanf(line.c_str(), "element vertex %d", &nvert);
-        sscanf(line.c_str(), "element face %d", &ntris);
-    } while (line != "end_header");
+    class SimplePlyParser : public PlyParserBase
+    {
+    public:
+        Mesh mesh;
 
-    if (nvert < 0 || ntris < 0) {
-        LOGTHROW(err2, std::runtime_error)
-                << filename << ": unknown PLY format.";
-    }
+        void loadHeader(const std::vector<PlyElement>& elements,
+                        const std::vector<std::string>& /* comments */) override
+        {
+            for (const auto& el : elements)
+            {
+                if (el.name == "vertex")
+                {
+                    mesh.vertices.reserve(el.count);
+
+                    // check vertex properties
+                    if (el.props.size() != 3)
+                    {
+                        LOGTHROW(err4, std::runtime_error)
+                            << "No additional vertex properties supported in "
+                               "simple parser";
+                    }
+                    utility::expect(el.props[0].name == "x");
+                    utility::expect(el.props[1].name == "y");
+                    utility::expect(el.props[2].name == "z");
+                }
+                else if (el.name == "face")
+                {
+                    mesh.faces.reserve(el.count);
+
+                    // check face properties
+                    if (el.props.size() != 1)
+                    {
+                        LOGTHROW(err4, std::runtime_error)
+                            << "No additional face properties supported in "
+                               "simple parser";
+                    }
+                    if (!ba::starts_with(el.props[0].type, "list"))
+                    {
+                        LOGTHROW(err4, std::runtime_error)
+                            << "Expected face property type to be a list, but "
+                               "got: "
+                            << el.props[0].type;
+                    }
+                    // does not check the name which should generally be
+                    // "vertex_index" but might vary
+                }
+                else
+                {
+                    LOGTHROW(err4, std::runtime_error)
+                        << "Unexpected element in header - not supported by "
+                           "simple parser: "
+                        << el.name;
+                }
+            }
+        }
 
-    Mesh mesh;
-    mesh.vertices.reserve(nvert);
+        void addVertex(std::istream& f) override
+        {
+            double x, y, z;
+            f >> x >> y >> z;
```

Comment:
We should check the state of the input stream. If the input file is corrupted; the operation might fail and there might be random/uninitialized values in `x,y,z`. If we haven't set exception throwing on the failure of operations on the input stream we must check the state manually.